router/CMakeLists.txt

```diff
@@ -35,4 +35,30 @@ set(router_SOURCES
 add_executable(skrouterd ${router_SOURCES})
 target_link_libraries(skrouterd skupper-router)
 
+if (ENABLE_PROFILE_GUIDED_OPTIMIZATION)
+    if("GNU" STREQUAL ${CMAKE_C_COMPILER_ID})
+        add_custom_command(TARGET skrouterd
+            POST_BUILD
+            # Run the tests with the profiled skrouterd binary
+            # Running the tests produce the files with the .gcda extension in the build/profile-dir folder
+            # For now, we are only running the adaptor based tests to keep it quick.
+            # Even if one of the adaptor tests fail, we will go ahead and create the final executable.
+            COMMAND ${CMAKE_CTEST_COMMAND} -j4 -V -R tcp || (exit 0)
+            COMMAND ${CMAKE_CTEST_COMMAND} -j4 -V -R http || (exit 0)
+            # Set the ENABLE_PROFILE_GUIDED_OPTIMIZATION=OFF so that we don't once again create profile data.
+            COMMAND ${CMAKE_COMMAND} ${CMAKE_BINARY_DIR} -DENABLE_PROFILE_GUIDED_OPTIMIZATION=OFF -DCMAKE_EXE_LINKER_FLAGS=-fprofile-use=${CMAKE_BINARY_DIR}/profile-data
+
+            # Feed the profile data we created from the above test runs using the -fprofile-use compiler option.
+            # Use the -fprofile-correction option to smoothen the counter updates when using multi-threaded applications.
+            COMMAND ${CMAKE_COMMAND} ${CMAKE_BINARY_DIR} -DC_EXTRA_GNU=-fprofile-use=${CMAKE_BINARY_DIR}/profile-data\\\;-fprofile-correction
```

Comment:
According to a random suse blog, you are suppose to use `-⁠fprofile-partial-training` at this point, if the tests were not entirely representative of the production load, which _might_(?) be the case.

> It is important that the train run exhibits the same characteristics as the real workload. Unless you use the option -⁠fprofile-partial-training in the second build, it needs to exercise the code that is also the most frequently executed in real use, otherwise it will be optimized for size and PGO would make more harm than good. With the option, GCC reverts to guessing properties of portions of the projects not exercised in the train run, as if they were compiled without profile feedback. This however also means that the code size will not perform better or shrink as much as one would expect from a PGO build.

https://documentation.suse.com/sbp/server-linux/single-html/SBP-GCC-11/index.html#sec-gcc11-pgo

Reply:
But the PGO build is currently running only a small portion of the tests, right? If you can be sure that "every function that needs to be fast" was executed during training, then the option is not helpful.

I don't insist on `-⁠fprofile-partial-training`, I'm only mentioning it as interesting possibility.

https://clearlinux.org/blogs-news/major-improvements-gcc-101

Reply:
> I will add the -fprofile-partial-training since it seems to only help.

It can hurt, in the articles they say if you really have some truly unused function, adding the option will make resulting binary bigger. I guess binary size is another thing to look for when experimenting, since one benefit of PGO can be smaller binaries.

But yeah, I guess I'd turn it on too, if it was my pr ;p

Reply:
```\\\;``` as the option separator? I did not try myself, but I think you need that, not just a `;`